packages/flutter/lib/src/widgets/router.dart

```diff
@@ -0,0 +1,1159 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// @dart = 2.8
+
+import 'dart:async';
+import 'dart:collection';
+import 'dart:convert';
+
+import 'package:flutter/foundation.dart';
+
+import 'basic.dart';
+import 'binding.dart';
+import 'framework.dart';
+import 'navigator.dart';
+import 'route_notification_messages.dart';
+
+/// A piece of routing information
+///
+/// This information flows two ways, from the [RouteInformationProvider] to the
+/// [Router] or from the [Router] to [RouteInformationProvider].
+///
+/// In the former case, the [RouteInformationProvider] notifies the [Router]
+/// widget when a new [RouteInformation] is available. The [Router] widget takes
+/// these information and navigates accordingly.
+///
+/// The latter case should only happen in a web application where the [Router]
+/// reports route change back to web engine.
+class RouteInformation {
+  /// Creates a route information.
+  const RouteInformation({this.routeName, this.state});
+
+  /// Creates a route information from a serialized string.
+  factory RouteInformation.fromSerialized(String serialized) {
+    final dynamic json = jsonDecode(serialized);
+    if (json is String)
+      return RouteInformation(routeName: json);
+    assert(json is Map<String, dynamic>);
+    final Map<String, dynamic> map = json as Map<String, dynamic>;
+    assert(map.containsKey('routeName') && map.containsKey('state'));
+    return RouteInformation(
+      routeName: json['routeName'] as String,
+      state: json['state'],
+    );
+  }
+
+  /// The location of the application.
+  ///
+  /// It is equivalent to the URL in a web application.
+  final String routeName;
```

Comment:
"name" implies this is just an identifier, but really it's a whole blob of data... "state" is also super generic... we should come up with better names here given the current design. I'll think about better names...

Reply:
that could work...